Improve dials.import_xds user interface

jbeilstenedmands opened the pull request:
I have tried to make this as permissive as possible to existing usage instructions.
The program now tries to create both experiment models and reflections files in one go.
More helpful error messages are printed if it can't find everything needed.

This works:
`dials.import_xds /path/to/folder/containing/xds/inp/`

also this:
`dials.import_xds /path/to/folder/containing/xds/inp/INTEGRATE.HKL`

also this:
`dials.import_xds /path/to/folder/containing/xds/inp/ INTEGRATE.HKL `

Specifying `SPOT.XDS` instead will pick up that (default is to pick INTEGRATE.HKL), and a few other more obscure uses.

These will also work if you're in an xds processing directory
`dials.import_xds .`
`dials.import_xds INTEGRATE.HKL`

Fixes #2647 
Fixes #2205

jbeilstenedmands commented:
I have just realised that this program is used in xia2, so I need to verify that still works correctly.

graeme-winter commented:
@jbeilstenedmands tested with some cubic insulin, integrated in P1 then fixed in CORRECT and got:

```console
Writing experiments to xds_models.expt
Reading /Users/graeme/data/necat-e-2025-04/2025-04-30-ins/insulin_04302025_1/xds/INTEGRATE.HKL
Read 1135737 reflections from /Users/graeme/data/necat-e-2025-04/2025-04-30-ins/insulin_04302025_1/xds/INTEGRATE.HKL
Undoing input variance model a, b = 4.0 0.0001
Reindex matrix:
-1 0 -1
0 1 1
1 1 0
Reindexing reflections
cctbx Internal Error: /Users/runner/miniforge3/conda-bld/cctbx-split_1735653491258/work/modules/cctbx_project/cctbx/sgtbx/rot_mx.cpp(36): Unsuitable value for rational rotation matrix.
```

because

```console

 ************ SELECTED SPACE GROUP AND UNIT CELL FOR THIS DATA SET ************

 SPACE_GROUP_NUMBER=  197
 UNIT_CELL_CONSTANTS=    78.77    78.77    78.77  90.000  90.000  90.000
 UNIT_CELL_A-AXIS=   -68.528    16.931    34.881
 UNIT_CELL_B-AXIS=    38.729    25.292    63.774
 UNIT_CELL_C-AXIS=     2.494    72.677   -30.298
 REFLECTIONS H,K,L FROM "INTEGRATE.HKL" WILL BE REINDEXED BY
 H' = ( -1*H +  0*K + -1*L)/ 1 +  0
 K' = (  0*H +  1*K +  1*L)/ 1 +  0
 L' = (  1*H +  1*K +  0*L)/ 1 +  0
```

This could pre-exist this PR tho

graeme-winter commented:
> @graeme-winter Looks like the same error that you reported here, so a preexisting issue I think: [#2647 (comment)](https://github.com/dials/dials/issues/2647#issuecomment-2058715586)

I am taking a few minutes to see if I can find a fix

graeme-winter commented:
> > @graeme-winter Looks like the same error that you reported here, so a preexisting issue I think: [#2647 (comment)](https://github.com/dials/dials/issues/2647#issuecomment-2058715586)
> 
> I am taking a few minutes to see if I can find a fix

A few minutes _did not_ arrive at a fix however did arrive at a question

Since we are importing data processed in XDS, with the matrix from XDS, how can the transformation to our system _not_ be a simple rotation 🤔

graeme-winter commented:
> > > @graeme-winter Looks like the same error that you reported here, so a preexisting issue I think: [#2647 (comment)](https://github.com/dials/dials/issues/2647#issuecomment-2058715586)
> > 
> > 
> > I am taking a few minutes to see if I can find a fix
> 
> A few minutes _did not_ arrive at a fix however did arrive at a question
> 
> Since we are importing data processed in XDS, with the matrix from XDS, how can the transformation to our system _not_ be a simple rotation 🤔

Oh, simples

```console
Read 1 experiment from /Users/graeme/data/necat-e-2025-04/2025-04-30-ins/insulin_04302025_1/xds/XDS_ASCII.HKL
```

This is where we went wrong

graeme-winter commented:
OK, looking some more

If we are reading from INTEGRATE.HKL for the reflections then we should take the matrix from e.g. XPARM.XDS not XDS_ASCII.HKL for consistency

This is not a trivial undertaking

jbeilstenedmands commented:
Ok, this is controllable by the `xds_file` parameter, if this is not specified we use the first entry in this list which exists:
`required_files_to_make_experiments = [
    "XDS_ASCII.HKL",
    "INTEGRATE.HKL",
    "GXPARM.XDS",
    "XPARM.XDS"
]`

Should the list order be different, perhaps `INTEGRATE.HKL` or `XPARM.XDS` first?

graeme-winter commented:
OK, for the record, with this one dumb change:

```python
required_files_to_make_experiments = [
#    "GXPARM.XDS",
    "XPARM.XDS",
    "INTEGRATE.HKL",
#    "XDS_ASCII.HKL",
]
```

It all works as expected - we should pair it up then & have _either_ XPARM + INTEGRATE _or_ GXPARM + XDS_ASCII

graeme-winter commented:
Also, for sake of sanity, we should have the experiment and reflection files have the same prefix as in the rest of DIALS ;-)

jbeilstenedmands commented:
Would it be sufficient to only ever try to read `XPARM.XDS` for the `xds_file` (or always try to read this first)? For `dials.import_xds`, I understand we are only supporting the use case of importing `SPOT.XDS` or `INTEGRATE.HKL`, so I don't understand why we would ever want to use `GXPARM.XDS` or `XDS_ASCII.HKL` for extra metadata? And you mention that `INTEGRATE.HKL` does not work due to not having detector distance?

jbeilstenedmands commented:
Seems like sometimes the detector distance is not picked up from `INTEGRATE.HKL`, but I think we should address this in the underlying library code here:
https://github.com/cctbx/cctbx_project/blob/b737ac62405e56f57e13267f28d0b11c4ae2c355/rstbx/cftbx/coordinate_frame_helpers.py#L306

Looks like the distance is typically given in this line in INTEGRATE.HKL:
`!ORGX=   1201.13  ORGY=   1212.74  DETECTOR_DISTANCE=   213.161`
so updating the library code to this is what we need, something like this:
```
if record.startswith('!ORGX='):
      ox = float(record.split()[1])
      oy = float(record.split()[3])
      distance = float(record.split()[5])
      continue
```

jbeilstenedmands commented:
This PR now depends on https://github.com/cctbx/cctbx_project/pull/1081

erikssod commented:
Apologies - I only came across this thread after it was closed. I was keeping an eye on this issue [here](https://github.com/dials/dials/issues/2205). 

I had a go and while `dials.import_xds` now indeed runs, I can't get any following steps to work. In the below example I'm trying to scale the data with following error. 

I'm on the overnight build for this `DIALS 3.dev.1327-g6ded2e84a Python 3.12.10`.

```
(base) de@MX3-DEV-DE2:.../fast_dp_results_thaumatin01_0$ dials.scale input.experiment=xds_models.expt input.reflections=integrate_hkl.refl

Traceback (most recent call last):
  File "/home/de/DIALS/dials-dev20250529/build/../modules/dials/src/dials/command_line/scale.py", line 259, in <module>
    run()
  File "/home/de/DIALS/dials-dev20250529/conda_base/lib/python3.12/contextlib.py", line 81, in inner
    return func(*args, **kwds)
           ^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/build/../modules/dials/src/dials/command_line/scale.py", line 214, in run
    params, options = parser.parse_args(args=args, show_diff_phil=False)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dials/src/dials/util/options.py", line 876, in parse_args
    params, args = self._phil_parser.parse_args(
                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dials/src/dials/util/options.py", line 496, in parse_args
    self._phil, unused = self.system_phil.fetch(
                         ^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/cctbx_project/libtbx/phil/__init__.py", line 1878, in fetch
    result_object = master_object.fetch(
                    ^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/cctbx_project/libtbx/phil/__init__.py", line 1904, in fetch
    candidate_as_str = master_object.extract_format(
                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/cctbx_project/libtbx/phil/__init__.py", line 1160, in extract_format
    return self.format(python_object=source.extract())
                                     ^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/cctbx_project/libtbx/phil/__init__.py", line 1145, in extract
    return self._type_from_words()(self.words, master=self)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dials/src/dials/util/phil.py", line 37, in from_words
    data=ExperimentListFactory.from_json_file(
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 899, in from_json_file
    return ExperimentListFactory.from_json(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 886, in from_json
    return ExperimentListFactory.from_dict(
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 876, in from_dict
    ).decode()
      ^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 492, in decode
    imageset = self._imageset_from_imageset_data(imageset_data, models)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/de/DIALS/dials-dev20250529/modules/dxtbx/src/dxtbx/model/experiment_list.py", line 412, in _imageset_from_imageset_data
    imageset.set_scan(scan)
RuntimeError: Please report this error at https://github.com/dials/dials/issues or to dials-user-group@jiscmail.ac.uk: 
dxtbx Internal Error: /scratch/jenkins_agent/workspace/dials_package_linux_conda3/build_dials/modules/dxtbx/src/dxtbx/imageset.h(365): DXTBX_ASSERT(index < scans_.size()) failure.
```

jbeilstenedmands commented:
Hi @erikssod , if you are able to send the xds.inp and integrate.hkl files to us then I will investigate, could you email them to the support list or me personally (james.beilsten-edmands at diamond.ac.uk). Thanks

erikssod commented:
Thanks! Files sent to your Diamond account.
